File: java/src/main/java/com/adaptionsoft/games/uglytrivia/Game.java
package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;
import java.util.LinkedList;

public class Game {
	public static final int NUMBER_OF_CELLS = 12;
	ArrayList<String> players = new ArrayList();
    int[] places = new int[6];
    int[] purses  = new int[6];
    boolean[] inPenaltyBox  = new boolean[6];
    
    LinkedList popQuestions = new LinkedList();
    LinkedList scienceQuestions = new LinkedList();
    LinkedList sportsQuestions = new LinkedList();
    LinkedList rockQuestions = new LinkedList();
    
    int currentPlayerIndex = 0;
    boolean isGettingOutOfPenaltyBox;
    
    public  Game(){
    	for (int i = 0; i < 50; i++) {
			popQuestions.addLast("Pop Question " + i);
			scienceQuestions.addLast(("Science Question " + i));
			sportsQuestions.addLast(("Sports Question " + i));
			rockQuestions.addLast(createRockQuestion(i));
    	}
    }

	public String createRockQuestion(int index){
		return "Rock Question " + index;
	}

	public void add(String playerName) {
		
		
	    players.add(playerName);
	    places[howManyPlayers()] = 0;
	    purses[howManyPlayers()] = 0;
	    inPenaltyBox[howManyPlayers()] = false;
	    
	    System.out.println(playerName + " was added");
	    System.out.println("They are player number " + players.size());

	}
	
	public int howManyPlayers() {
		return players.size();
	}

	public void roll(int roll) {
		System.out.println(currentPlayerName() + " is the current player");
		System.out.println("They have rolled a " + roll);
		
		if (inPenaltyBox[currentPlayerIndex]) {
			if (roll % 2 != 0) {
				isGettingOutOfPenaltyBox = true;
				
				System.out.println(currentPlayerName() + " is getting out of the penalty box");
				places[currentPlayerIndex] = places[currentPlayerIndex] + roll;
				if (places[currentPlayerIndex] >= NUMBER_OF_CELLS) places[currentPlayerIndex] = places[currentPlayerIndex] - NUMBER_OF_CELLS;
				
				System.out.println(currentPlayerName()
						+ "'s new location is " 
						+ places[currentPlayerIndex]);
				System.out.println("The category is " + currentCategory());
				askQuestion();
			} else {
				System.out.println(currentPlayerName() + " is not getting out of the penalty box");
				isGettingOutOfPenaltyBox = false;
				}
			
		} else {
		
			places[currentPlayerIndex] = places[currentPlayerIndex] + roll;
			if (places[currentPlayerIndex] >= NUMBER_OF_CELLS) places[currentPlayerIndex] = places[currentPlayerIndex] - NUMBER_OF_CELLS;
			
			System.out.println(currentPlayerName()
					+ "'s new location is " 
					+ places[currentPlayerIndex]);
			System.out.println("The category is " + currentCategory());
			askQuestion();
		}
		
	}

	private void askQuestion() {
		if (currentCategory() == Category.POP)
			System.out.println(popQuestions.removeFirst());
		if (currentCategory() == Category.SCIENCE)
			System.out.println(scienceQuestions.removeFirst());
		if (currentCategory() == Category.SPORTS)
			System.out.println(sportsQuestions.removeFirst());
		if (currentCategory() == Category.ROCK)
			System.out.println(rockQuestions.removeFirst());		
	}
	
	
	private Category currentCategory() {
		if (places[currentPlayerIndex] == 0) return Category.POP;
		if (places[currentPlayerIndex] == 4) return Category.POP;
		if (places[currentPlayerIndex] == 8) return Category.POP;
		if (places[currentPlayerIndex] == 1) return Category.SCIENCE;
		if (places[currentPlayerIndex] == 5) return Category.SCIENCE;
		if (places[currentPlayerIndex] == 9) return Category.SCIENCE;
		if (places[currentPlayerIndex] == 2) return Category.SPORTS;
		if (places[currentPlayerIndex] == 6) return Category.SPORTS;
		if (places[currentPlayerIndex] == 10) return Category.SPORTS;
		return Category.ROCK;
	}

	public boolean wasCorrectlyAnswered() {
		if (inPenaltyBox[currentPlayerIndex]){
			if (isGettingOutOfPenaltyBox) {
				System.out.println("Answer was correct!!!!");
				purses[currentPlayerIndex]++;
				System.out.println(currentPlayerName()
						+ " now has "
						+ purses[currentPlayerIndex]
						+ " Gold Coins.");
				
				boolean winner = didPlayerWin();
				currentPlayerIndex++;
				if (currentPlayerIndex == players.size()) currentPlayerIndex = 0;
				
				return winner;
			} else {
				currentPlayerIndex++;
				if (currentPlayerIndex == players.size()) currentPlayerIndex = 0;
				return true;
			}
			
			
			
		} else {
		
			System.out.println("Answer was corrent!!!!");
			purses[currentPlayerIndex]++;
			System.out.println(currentPlayerName()
					+ " now has "
					+ purses[currentPlayerIndex]
					+ " Gold Coins.");
			
			boolean winner = didPlayerWin();
			currentPlayerIndex++;
			if (currentPlayerIndex == players.size()) currentPlayerIndex = 0;
			
			return winner;
		}
	}

	private String currentPlayerName() {
		return players.get(currentPlayerIndex);
	}

	public boolean wrongAnswer(){
		System.out.println("Question was incorrectly answered");
		System.out.println(currentPlayerName() + " was sent to the penalty box");
		inPenaltyBox[currentPlayerIndex] = true;
		
		currentPlayerIndex++;
		if (currentPlayerIndex == players.size()) currentPlayerIndex = 0;
		return true;
	}


	private boolean didPlayerWin() {
		return !(purses[currentPlayerIndex] == 6);
	}
}

File: swift/Trivia/Game.swift
//
//  Game.swift
//  Trivia
//
//  Created by Oliver Eikemeier on 13.10.15.
//  Copyright © 2015 Legacy Coderetreat. All rights reserved.
//

import Foundation

public class Game {
    var players = [String]()
    var places = [Int](repeating: 0, count: 6)
    var purses  = [Int](repeating: 0, count: 6)
    var inPenaltyBox  = [Bool](repeating: false, count: 6)
    
    var popQuestions = [String]()
    var scienceQuestions = [String]()
    var sportsQuestions = [String]()
    var rockQuestions = [String]()
    
    var currentPlayer = 0
    var isGettingOutOfPenaltyBox = false
    
    private var printer: Printer = ConsolePrinter()
    
    public convenience init(printer: Printer){
        self.init()
        self.printer = printer
    }
    
    public  init(){
    	for i in 0..<50 {
			popQuestions.append("Pop Question \(i)")
			scienceQuestions.append(("Science Question \(i)"))
			sportsQuestions.append(("Sports Question \(i)"))
            rockQuestions.append(createRockQuestion(index: i))
    	}
    }

    public func createRockQuestion(index: Int) -> String{
		return "Rock Question \(index)"
	}
	
	public func isPlayable() -> Bool {
		return howManyPlayers() >= 2
	}

    public func add(playerName: String) -> Bool {
		
		
	    players.append(playerName)
	    places[howManyPlayers()] = 0
	    purses[howManyPlayers()] = 0
	    inPenaltyBox[howManyPlayers()] = false
	    
	    printer.output(playerName, "was added")
	    printer.output("They are player number", players.count)
		return true
	}
	
	public func howManyPlayers() -> Int {
		return players.count
	}

    public func roll(roll: Int) {
		printer.output(players[currentPlayer], "is the current player")
		printer.output("They have rolled a", roll)
		
		if inPenaltyBox[currentPlayer] {
			if roll % 2 != 0 {
				isGettingOutOfPenaltyBox = true
				
				printer.output(players[currentPlayer], "is getting out of the penalty box")
				places[currentPlayer] = places[currentPlayer] + roll
                if places[currentPlayer] > 11 {places[currentPlayer] = places[currentPlayer] - 12}
				
				printer.output(players[currentPlayer]
						+ "'s new location is",
						places[currentPlayer])
				printer.output("The category is", currentCategory())
				askQuestion()
			} else {
				printer.output(players[currentPlayer], "is not getting out of the penalty box")
				isGettingOutOfPenaltyBox = false
				}
			
		} else {
		
			places[currentPlayer] = places[currentPlayer] + roll
            if places[currentPlayer] > 11 {places[currentPlayer] = places[currentPlayer] - 12}
			
			printer.output(players[currentPlayer]
					+ "'s new location is",
					places[currentPlayer])
			printer.output("The category is", currentCategory())
			askQuestion()
		}
		
	}

 	private func askQuestion() {
        if currentCategory() == "Pop" {
            printer.output(popQuestions.removeFirst())}
        if currentCategory() == "Science"{
            printer.output(scienceQuestions.removeFirst())}
        if currentCategory() == "Sports"{
            printer.output(sportsQuestions.removeFirst())}
        if currentCategory() == "Rock"{
            printer.output(rockQuestions.removeFirst())}
	}
	
	
	private func currentCategory() -> String {
        if places[currentPlayer] == 0 {return "Pop"}
		if places[currentPlayer] == 4 {return "Pop"}
		if places[currentPlayer] == 8 {return "Pop"}
		if places[currentPlayer] == 1 {return "Science"}
		if places[currentPlayer] == 5 {return "Science"}
		if places[currentPlayer] == 9 {return "Science"}
		if places[currentPlayer] == 2 {return "Sports"}
		if places[currentPlayer] == 6 {return "Sports"}
		if places[currentPlayer] == 10 {return "Sports"}
		return "Rock"
	}

	public func wasCorrectlyAnswered() -> Bool {
		if inPenaltyBox[currentPlayer]{
			if isGettingOutOfPenaltyBox {
				printer.output("Answer was correct!!!!")
				purses[currentPlayer] += 1
				printer.output(players[currentPlayer],
						"now has",
						purses[currentPlayer],
						"Gold Coins.")
				
				let winner = didPlayerWin
				currentPlayer += 1
                if currentPlayer == players.count {currentPlayer = 0}
				
				return winner
			} else {
				currentPlayer += 1
                if currentPlayer == players.count {currentPlayer = 0}
				return true
			}
			
			
			
		} else {
		
			printer.output("Answer was corrent!!!!")
			purses[currentPlayer] += 1
			printer.output(players[currentPlayer],
					"now has",
                    purses[currentPlayer],
					"Gold Coins.")
			
			let winner = didPlayerWin
			currentPlayer += 1
            if currentPlayer == players.count {currentPlayer = 0}
			
			return winner
		}
	}
	
	public func wrongAnswer()->Bool{
        printer.output("Question was incorrectly answered")
		printer.output(players[currentPlayer], "was sent to the penalty box")
		inPenaltyBox[currentPlayer] = true
		
		currentPlayer += 1
        if currentPlayer == players.count {currentPlayer = 0}
		return true
	}


    private var didPlayerWin: Bool {
		return !(purses[currentPlayer] == 6)
	}
}
